pkg/factory/factory.go

```diff
@@ -0,0 +1,174 @@
+		}
+	}
+	return instance
+}
+
+// GetResourceServer returns an instance of ResourceServer for a ResourcePool
+func (rf *resourceFactory) GetResourceServer(rp types.ResourcePool) (types.ResourceServer, error) {
+	if rp != nil {
+		prefix := rf.endPointPrefix
+		if prefixOverride := rp.GetResourcePrefix(); prefixOverride != "" {
+			prefix = prefixOverride
+		}
+		return resources.NewResourceServer(prefix, rf.endPointSuffix, rf.pluginWatch, rp), nil
+	}
+	return nil, fmt.Errorf("factory: unable to get resource pool object")
+}
+
+// GetInfoProvider returns an instance of DeviceInfoProvider using name as string
+func (rf *resourceFactory) GetInfoProvider(name string) types.DeviceInfoProvider {
+	switch name {
+	case "vfio-pci":
+		return resources.NewVfioResource()
+	case "uio", "igb_uio":
+		return resources.NewUioResource()
+	default:
+		return resources.NewGenericResource()
+	}
+}
+
+// GetSelector returns an instance of DeviceSelector using selector attribute string and its associated values
+func (rf *resourceFactory) GetSelector(attr string, values []string) (types.DeviceSelector, error) {
+	switch attr {
+	case "vendors":
+		return resources.NewVendorSelector(values), nil
+	case "devices":
+		return resources.NewDeviceSelector(values), nil
+	case "drivers":
+		return resources.NewDriverSelector(values), nil
+	case "pfNames":
+		return resources.NewPfNameSelector(values), nil
+	case "linkTypes":
+		return resources.NewLinkTypeSelector(values), nil
+	case "ddpProfiles":
+		return resources.NewDdpSelector(values), nil
+	default:
+		return nil, fmt.Errorf("GetSelector(): invalid attribute %s", attr)
+	}
+}
+
+// GetResourcePool returns an instance of resourcePool
+func (rf *resourceFactory) GetResourcePool(rc *types.ResourceConfig, filteredDevice []types.PciDevice) (types.ResourcePool, error) {
+
+	devicePool := make(map[string]types.PciDevice, 0)
+	apiDevices := make(map[string]*pluginapi.Device)
+	for _, dev := range filteredDevice {
+		pciAddr := dev.GetPciAddr()
+		devicePool[pciAddr] = dev
+		apiDevices[pciAddr] = dev.GetAPIDevice()
+		glog.Infof("device added: [pciAddr: %s, vendor: %s, device: %s, driver: %s]",
+			dev.GetPciAddr(),
+			dev.GetVendor(),
+			dev.GetDeviceCode(),
+			dev.GetDriver())
+	}
+
+	var rPool types.ResourcePool
+	var err error
+	switch rc.DeviceType {
+	case types.NetDeviceType:
+		if len(filteredDevice) > 0 {
+			if _, ok := filteredDevice[0].(types.PciNetDevice); ok {
+				rPool = netdevice.NewNetResourcePool(rc, apiDevices, devicePool)
+			} else {
+				rPool = nil
+				err = fmt.Errorf("invalid device list for NetDeviceType")
+			}
+		}
+	case types.AcceleratorType:
+		if len(filteredDevice) > 0 {
+			if _, ok := filteredDevice[0].(types.AccelDevice); ok {
+				rPool = accelerator.NewAccelResourcePool(rc, apiDevices, devicePool)
+			} else {
+				rPool = nil
+				err = fmt.Errorf("invalid device list for AcceleratorType")
+			}
+		}
+	default:
+		rPool = resources.NewResourcePool(rc, apiDevices, devicePool)
```

Comment:
any thoughts?

Reply:
Sorry, I've missed this comment before.
The NewResourcePool will return a generic pool implementation, we expect not to reach here. I added this here just in case in future we have other devices that may work with a generic pool. 
For AcceleratorType, we could have used the generic pool but I think it's good idea to use it's own implementation for future extension.

We could change 'default' to return nil and error with not supporting device.